src/pymor/algorithms/cholesky_qr.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+
+import numpy as np
+import scipy.linalg as spla
+
+from pymor.core.logger import getLogger
+
+logger = getLogger('pymor.algorithms.cholesky_qr.cholesky_qr')
+
+
+def _chol_qr(V, product=None, gramian=None, check_finite=True):
+    X = V.gramian(product=product) if gramian is None else gramian
+
+    while True:
+        try:
+            R = spla.cholesky(X, overwrite_a=True, check_finite=check_finite)
+            break
+        except spla.LinAlgError:
+            logger.warning('Cholesky factorization broke down! Matrix is ill-conditioned.')
+            m, n = V.dim, len(V)
+            eps = 7. / 3 - 4. / 3 - 1
+            s = 11*(m*n + n*(n+1))*eps*spla.norm(V.to_numpy().T, ord=2, check_finite=check_finite)
+            logger.info(f'Applying shift: {s}')
+            np.fill_diagonal(X, np.diag(X)+s)
+
+    Q = spla.solve_triangular(R, V.to_numpy(), trans='T', overwrite_b=True, check_finite=check_finite)
```

Comment:
Here you convert the entire input array to a numpy array, which significantly limits the applicability of the algortithm.  This could completely avoided if one was willing to compute the inverse of `R`. I don't know what that would mean for the stability of the algorithm, though ..

Reply:
Good news, I've inverted `R` with `scipy.linalg.lapack.dtrtri` which runs at the same speed but is about 30% more accurate in my tests.